sprinkle/user-facing/pytorch_trial.md

```diff
@@ -0,0 +1,240 @@
+* [x] Optionally you can override the your trial's data loaders with arguments
+      to `context.fit()`
+
+## Training Example
+
+Here is an example training script, which can either be submitted to the
+cluster or it can be invoked to train locally right on your laptop.
+
+```python
+# train.py
+"""
+trains a trivial model to briefly show the PyTorchTrial API
+and also the new proposed Trainer API.
+"""
+
+import torch
+from torch import nn
+
+import determined as det
+import determined.pytorch
+
+## Data Definition ##
+
+class OnesDataset(torch.utils.data.Dataset):
+    def __len__(self):
+        return 64
+
+    def __getitem__(self, index):
+        # return data, labels
+        return torch.Tensor([1.0]), torch.Tensor([1.0])
+
+## Model Definition ##
+
+class OneVarPytorchTrial(det.pytorch.PyTorchTrial):
+    def __init__(self, context):
+        # Watch out! This is a trial context, not a trainer context like below!
+        self.context = context
+
+        # Wrap any models with context.wrap_model()
+        self.model = context.wrap_model(nn.Linear(1, 1, False))
+
+        lr = context.hparams["learning_rate"]
+
+        # Wrap any optimizers with context.wrap_optimizer()
+        self.opt = context.wrap_optimizer(
+            torch.optim.SGD(self.model.parameters(), lr=lr)
+        )
+
+    def train_batch(self, batch, epoch_idx, batch_idx):
+        """
+        Tell us how to calculate the loss/metrics for a single batch.
+        """
+        data, labels = batch
+        preds = self.model(batch)
+        loss = torch.nn.MSELoss()(preds, labels)
+        # use context.backward() and context.step_optimizer()
+        self.context.backward(loss)
+        self.context.step_optimizer(self.opt)
+        return {"loss": loss}
+
+    def evaluate_batch(self, batch: pytorch.TorchData):
+        """
+        Tell us how to calculate validation metrics for a single batch.
+        """
+        data, labels = batch
+        preds = self.model(data)
+        loss = torch.nn.MSELoss()(preds, labels)
+        return {"loss": loss}
+
+    # you can also define build_training_data_loader() or
+    # build_validation_data_loader() as methods of your PyTorchTrial,
+    # but in this example we show how to pass the data loaders in via
+    # the trainer API.
+
+## Training script ##
+
+if __name__ == "__main__":
+
+    # Watch out! This is a trainer context, not a trial context like above!
+    # (this step will also initialize random seeds automatically)
+    context = det.pytorch.init()
+
+    # Access your hparams from the trainer context.
+    hparams = context.training.hparams
+    global_batch_size = context.training.hparams["global_batch_size"]
+
+    # Optional for distributed training: cacluate this worker's
+    # batch_size so that the effective batch_size across all workers
+    # remains constant.  If you choose not to use this, the
+    # effective batch_size will be multiplied by the number slots_per_trial.
```

Comment:
wait, so are we changing what global_batch_size means now?

Reply:
yes, because `global_batch_size` wouldn't mean anything to us except in the backwards compatible case.  The user gets to do what they want with their `batch_size`, and we just make it easy for them to do what they want.

Reply:
we could rename `context.distributed.shard_batch_size` as `context.per_slot_batch_size(global_batch_size)` where the input parameter can be skipped for backwards compatibility, if you set `global_batch_size` as a hyperparameter?